Extension/src/Utility/Text/taggedLiteral.ts

```diff
@@ -0,0 +1,254 @@
+                        continue; // ignore whitespace
+
+                    case ':':
+                    case '.':
+                        expression += ':';
+                        continue;
+                }
+                if (expression) {
+                    if (isIdentifierPart(char.codePointAt(0)!) || char === '-' || char === '/') {
+                        expression += char;
+                        continue;
+                    }
+                    // error, fall thru
+                } else if (isIdentifierStart(char.codePointAt(0)!) || char === '-' || char === '/') {
+                    expression += char;
+                    continue;
+                }
+
+                // not a valid character for an expression
+                result.state = 'error';
+                result.message = `Unexpected character '${char}' in expression ${expression}`;
+                return result;
+        }
+    }
+
+    switch (result.state) {
+        case 'escape':
+            result.state = 'error';
+            result.message = 'Unexpected end of string (trailing backslash)';
+            return result;
+        case 'substitution':
+            result.state = 'error';
+            result.message = 'Unexpected end of string parsing expression ${ ';
+            return result;
+        case 'dollar':
+            template += '$';
+            break;
+    }
+    result.state = 'ok';
+    result.template.push(template);
+    return result;
+}
+
+function split(expression: string) {
+    return (expression.match(/(.*?):(.*)/) || ['', '', expression]).slice(1);
+}
+
+function resolveValue(expression: string, context: Record<string, any>, customResolver = (_prefix: string, _expression: string) => ''): string {
+    const [prefix, suffix] = split(expression);
+
+    function joinIfArray(value: any, separator = '\u0007') {
+        return Array.isArray(value) ? value.join(separator) : value.toString();
+    }
+
+    if (prefix) {
+        const variable = context[prefix];
+        if (variable !== undefined && variable !== null) { // did we get back an actual value
+            // its a child of a variable
+            return joinIfArray(suffix.includes(':') ? // is the suffix another expression?
+                resolveValue(suffix, variable) : // Yeah, resolve it
+                variable[suffix] ?? customResolver(prefix, suffix) ?? ''); // No, return the member of the variable, or dynamic, or empty string
+        }
+
+        // no variable by that name, so return the dynamic value, or an empty string
+        return joinIfArray(customResolver(prefix, suffix) ?? '');
+    }
+
+    // look up the value in the variables, or ask the dynamic function to resolve it, failing that, an empty string
+    return joinIfArray(context[suffix] ?? customResolver(prefix, suffix) ?? '');
+}
+
+// eslint-disable-next-line @typescript-eslint/naming-convention
+class as {
+    /** returns the value as a number (NOT NaN) or undefined */

```

Comment:
Is `NOT` supposed to be in all caps in this comment and the 2 below?

Reply:
Yes. It is for emphasis.